File: C/wallpaper.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

/*
 * wallpaper.c
 *
 * Usage gcc wallaper.c -o wallpaper 
 * run ./wallpaper 
 * Version: 1.0.3 - 
 *
 * Copyright (c) 2023 Jeremy Stevens
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
 
#define WALLPAPER_VERSION "1.0.3"
#define AUTHOR_NAME "Jeremy Stevens"

// Color codes for the text
#define HEADER "\033[95m"
#define OKBLUE "\033[94m"
#define OKGREEN "\033[92m"
#define WARNING "\033[93m"
#define FAIL "\033[91m"
#define ENDC "\033[0m"
#define BOLD "\033[1m"
#define UNDERLINE "\033[4m"

#define MAX_INPUT_LENGTH 100

#ifdef _WIN32
#include <windows.h>
#include <Lmcons.h>
#else
#include <sys/stat.h>
#include <unistd.h>
#endif

bool directoryExists(const char *path) {
    struct stat statbuf;
    return (stat(path, &statbuf) == 0 && S_ISDIR(statbuf.st_mode));
}

void createDirectory(const char *path) {
#ifdef _WIN32
    CreateDirectory(path, NULL);
#else
    mkdir(path, 0700); // Read, write, execute permissions for the user
#endif
}

void getUserInput(char *input, int length) {
    printf("Please enter your input: ");
    if (fgets(input, length, stdin) != NULL) {
        size_t len = strlen(input);
        if (len > 0 && input[len - 1] == '\n') {
            input[len - 1] = '\0';
        }
    } else {
        fprintf(stderr, "Error reading input.\n");
    }
}

static char picturesPath[256]; // Static variable to store the pictures path

int initializePicturesDirectory() {
#ifdef _WIN32
    char username[UNLEN + 1];
    DWORD username_len = UNLEN + 1;
    GetUserName(username, &username_len);
    snprintf(picturesPath, sizeof(picturesPath), "C:\\Users\\%s\\Pictures", username);
#else
    const char *homeDir = getenv("HOME");
    if (homeDir == NULL) {
        fprintf(stderr, "Unable to find home directory.\n");
        return 1;
    }
    snprintf(picturesPath, sizeof(picturesPath), "%s/Pictures", homeDir);
#endif

    if (!directoryExists(picturesPath)) {
        printf("The Pictures directory does not exist. Creating one...\n");
        createDirectory(picturesPath);
        if (!directoryExists(picturesPath)) {
            fprintf(stderr, "Failed to create Pictures directory.\n");
            return 1;
        }
        printf("Pictures directory created successfully.\n");
    } else {
        printf(OKBLUE "Found Pictures directory: %s\n" ENDC, picturesPath);
    }

    char downloadHistoryPath[256];
    snprintf(downloadHistoryPath, sizeof(downloadHistoryPath), "%s/download_history.txt", picturesPath);
    FILE *downloadHistoryFile = fopen(downloadHistoryPath, "w");
    if (downloadHistoryFile != NULL) {
        printf("Download history file created successfully.\n");
        fclose(downloadHistoryFile);
    } else {
        fprintf(stderr, "Failed to create download history file.\n");
        return 1;
    }

    return 0; // Success
}

void printAppInfo() {
    printf(OKGREEN "Wallpaper v%s\n" ENDC, WALLPAPER_VERSION);
    printf(HEADER "Created by %s\n" ENDC, AUTHOR_NAME);
}

void clearConsole() {
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

int initializeApp() {
    return initializePicturesDirectory();
}

void getUserPreferences(char *resolution, char *type, char *keyword, int *number, int *sleepDuration) {
    printf("Enter desired wallpaper resolution in 1920x1080 format: ");
    getUserInput(resolution, MAX_INPUT_LENGTH);

    printf("Enter your desired wallpaper type (keyword/random): ");
    getUserInput(type, MAX_INPUT_LENGTH);

    if (strcmp(type, "keyword") == 0) {
        printf("Please enter your desired wallpaper keyword: ");
        getUserInput(keyword, MAX_INPUT_LENGTH);
    } else {
        strcpy(keyword, "");
    }

    char numberInput[MAX_INPUT_LENGTH];
    printf("Enter the number of wallpapers you want to download: ");
    getUserInput(numberInput, MAX_INPUT_LENGTH);
    *number = atoi(numberInput);

    char sleepInput[MAX_INPUT_LENGTH];
    printf("enter time to sleep between downloads in seconds: ");
    getUserInput(sleepInput, MAX_INPUT_LENGTH);
    *sleepDuration = atoi(sleepInput);
}

void downloadWallpaper(int *number, int sleepDuration, const char *type, const char *keyword, const char *path, const char *resolution) {
    char downloadHistoryPath[256];
    snprintf(downloadHistoryPath, sizeof(downloadHistoryPath), "%s/download_history.txt", path);
    FILE *downloadHistoryFile = fopen(downloadHistoryPath, "a+");

    if (downloadHistoryFile == NULL) {
        fprintf(stderr, "Failed to open download history file.\n");
        return;
    }

    char line[1024];
    int downloadedCount = 0;
    while (downloadedCount < *number) {
        char url[256];
        char imageName[256];
        char command[512];
        char hashCommand[512];
        char hashValue[128];

        // Generate URL based on type
        if (strcmp(type, "keyword") == 0) {
            snprintf(url, sizeof(url), "https://source.unsplash.com/featured/%s/?%s", resolution, keyword);
        } else {
            snprintf(url, sizeof(url), "https://source.unsplash.com/random/%s", resolution);
        }

        // Generate a unique name for the image
        snprintf(imageName, sizeof(imageName), "%s/wallpaper_%d.jpg", path, downloadedCount + 1);

        // Download the image using curl with -L to follow redirects (must have the -L flag)
        snprintf(command, sizeof(command), "curl -L -o \"%s\" \"%s\"", imageName, url);
        int result = system(command);

        if (result != 0) {
            fprintf(stderr, "Error occurred during downloading. Exit code: %d\n", result);
            continue; // Skip further processing for this download
        }

        // Generate a unique identifier for the image (hash)
        snprintf(hashCommand, sizeof(hashCommand), "md5sum \"%s\" | cut -d ' ' -f 1", imageName);
        FILE *hashPipe = popen(hashCommand, "r");
        if (hashPipe == NULL) {
            fprintf(stderr, "Failed to compute hash for the image.\n");
            remove(imageName); // Remove the downloaded image
            break;
        }
        fgets(hashValue, sizeof(hashValue), hashPipe);
        pclose(hashPipe);

        // Check if the hash is already in download history
        bool alreadyDownloaded = false;
        rewind(downloadHistoryFile);
        while (fgets(line, sizeof(line), downloadHistoryFile)) {
            if (strstr(line, hashValue) != NULL) {
                alreadyDownloaded = true;
                break;
            }
        }

        if (!alreadyDownloaded) {
            // Add the hash to download history
            fprintf(downloadHistoryFile, "%s\n", hashValue);
            fflush(downloadHistoryFile);

            printf("Downloaded wallpaper %d/%d\n", downloadedCount + 1, *number);
            downloadedCount++;

            // Sleep between downloads
            sleep(sleepDuration);
        } else {
            printf("Skipping already downloaded image.\n");
            remove(imageName); // Remove the duplicate image
        }
    }

    fclose(downloadHistoryFile);
}


void performWallpaperDownload(const char *resolution, const char *type, const char *keyword, int number, int sleepDuration) {
    time_t start = time(NULL);
    downloadWallpaper(&number, sleepDuration, type, keyword, picturesPath, resolution);
    time_t end = time(NULL);
    printf("Time taken: %.2fs\n", difftime(end, start));
}

int main() {
    printAppInfo();
    sleep(2);
    clearConsole();

    if (initializeApp() != 0) {
        fprintf(stderr, "Failed to initialize application.\n");
        return 1;
    }

    char resolution[MAX_INPUT_LENGTH];
    char type[MAX_INPUT_LENGTH];
    char keyword[MAX_INPUT_LENGTH];
    int number, sleepDuration;

    getUserPreferences(resolution, type, keyword, &number, &sleepDuration);
    performWallpaperDownload(resolution, type, keyword, number, sleepDuration);

    return 0;
}

File: C#/wallpaper.cs
using System;
using System.IO;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

/*
 * Copyright (C) 2023  Jeremy Stevens
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

class Program
{
    static async Task Main(string[] args)
    {
        // Display the banner
        Console.WriteLine(@"
 █░█░█▀█░█░░░█░░░█▀█░█▀█░█▀█░█▀▀░█▀▄░░░█▀▄░█▀█░█░█░█▀█░█░░░█▀█░█▀█░█▀▄░█▀▀░█▀▄
░█▄█░█▀█░█░░░█░░░█▀▀░█▀█░█▀▀░█▀▀░█▀▄░░░█░█░█░█░█▄█░█░█░█░░░█░█░█▀█░█░█░█▀▀░█▀▄
░▀░▀░▀░▀░▀▀▀░▀▀▀░▀░░░▀░▀░▀░░░▀▀▀░▀░▀░░░▀▀░░▀▀▀░▀░▀░▀░▀░▀▀▀░▀▀▀░▀░▀░▀▀░░▀▀▀░▀░▀
Rev 3.5 by Jeremy Stevens
");

        // Get user input for screen resolution
        Console.Write("Enter your screen resolution (e.g., 1920x1080): ");
        string desiredResolution = Console.ReadLine();

        // Get user input for download mode (random or keyword)
        Console.Write("Choose the download mode (random/keyword): ");
        string downloadMode = Console.ReadLine();

        string keyword = null;

        if (downloadMode == "keyword")
        {
            Console.Write("Enter a keyword for wallpapers: ");
            keyword = Console.ReadLine();
        }

        Console.Write("Enter the maximum number of wallpapers to download (default: 100): ");
        if (!int.TryParse(Console.ReadLine(), out int maxWallpapers))
        {
            maxWallpapers = 100;
        }

        Console.Write("Enter the sleep duration in seconds between downloading wallpapers (default: 300): ");
        if (!int.TryParse(Console.ReadLine(), out int sleepDuration))
        {
            sleepDuration = 300;
        }

        // Get the current logged-in user
        string user = Environment.UserName;

        // Define the directory where wallpapers will be stored
        string wallpaperDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Pictures");

        // Path to the download history file
        string downloadHistoryFile = Path.Combine(wallpaperDir, "download_history.txt");

        // Set the initial value of the downloaded wallpaper counter to 0
        int downloadedWallpapers = 0;

        // Create the directory if it doesn't exist
        if (!Directory.Exists(wallpaperDir))
        {
            Directory.CreateDirectory(wallpaperDir);
        }

        // Create the download history file if it doesn't exist
        if (!File.Exists(downloadHistoryFile))
        {
            File.Create(downloadHistoryFile).Close();
        }

        while (downloadedWallpapers < maxWallpapers)
        {
            // Download the wallpaper based on the chosen mode (random or keyword)
            string imageUrl = downloadMode == "random"
                ? await GetRandomImage(desiredResolution)
                : await GetKeywordImage(desiredResolution, keyword);

            // Calculate the MD5 hash of the image
            string imageHash = CalculateMd5Hash(await DownloadImage(imageUrl));

            // Check if the image hash is already in the download history
            if (!IsImageAlreadyDownloaded(downloadHistoryFile, imageHash))
            {
                // Save the image to the wallpaper directory
                string uniqueFilename = $"wallpaper_{DateTime.Now:yyyyMMddHHmmss}.jpg";
                string imagePath = Path.Combine(wallpaperDir, uniqueFilename);
                await SaveImage(imageUrl, imagePath);

                // Append the image hash to the download history file
                File.AppendAllText(downloadHistoryFile, imageHash + Environment.NewLine);

                // Increment the downloaded wallpaper counter
                downloadedWallpapers++;

                Console.WriteLine($"Downloaded wallpaper {downloadedWallpapers} of {maxWallpapers}");
            }

            // Sleep for the specified duration before downloading the next wallpaper
            Thread.Sleep(sleepDuration * 1000);
        }

        Console.WriteLine($"Maximum number of wallpapers ({maxWallpapers}) downloaded. Exiting...");
    }

    static async Task<string> GetRandomImage(string resolution)
    {
        string imageUrl = $"https://source.unsplash.com/random/{resolution}";
        return await GetFinalImageUrl(imageUrl);
    }

    static async Task<string> GetKeywordImage(string resolution, string keyword)
    {
        string imageUrl = $"https://source.unsplash.com/featured/{resolution}/?{keyword}";
        return await GetFinalImageUrl(imageUrl);
    }

    static async Task<string> GetFinalImageUrl(string imageUrl)
    {
        using (HttpClient httpClient = new HttpClient())
        {
            HttpResponseMessage response = await httpClient.GetAsync(imageUrl, HttpCompletionOption.ResponseHeadersRead);
            return response.RequestMessage.RequestUri.ToString();
        }
    }

    static async Task<byte[]> DownloadImage(string imageUrl)
    {
        using (HttpClient httpClient = new HttpClient())
        {
            return await httpClient.GetByteArrayAsync(imageUrl);
        }
    }

    static string CalculateMd5Hash(byte[] data)
    {
        using (MD5 md5 = MD5.Create())
        {
            byte[] hash = md5.ComputeHash(data);
            StringBuilder hashBuilder = new StringBuilder();
            for (int i = 0; i < hash.Length; i++)
            {
                hashBuilder.Append(hash[i].ToString("x2"));
            }
            return hashBuilder.ToString();
        }
    }

    static bool IsImageAlreadyDownloaded(string downloadHistoryFile, string imageHash)
    {
        if (!File.Exists(downloadHistoryFile))
        {
            return false;
        }

        string[] lines = File.ReadAllLines(downloadHistoryFile);
        return Array.Exists(lines, line => line == imageHash);
    }

    static async Task SaveImage(string imageUrl, string imagePath)
    {
        using (HttpClient httpClient = new HttpClient())
        {
            byte[] imageData = await httpClient.GetByteArrayAsync(imageUrl);
            File.WriteAllBytes(imagePath, imageData);
        }
    }
}
